---
title: "Linear models in R"
output: 
    html_document:
        theme: cerulean
        css: style.css
---

## Workshop notes

* [Slideshow](slides/linear_thinking.html)
* [Linear models in R](topics/linear_models.html)

## Setup

This workshop is designed to work with [RStudio Cloud](https://rstudio.cloud/). Go to https://rstudio.cloud/ (Monash users can log in with their Monash google account) and create a new project. The workshop can also be done using R locally on your laptop (if doing this, we also recommend you create a new project to contain the files).

Running the R code below will download files and install packages used in this workshop.

```{r eval=FALSE}
# Download data
download.file(
  "https://monashdatafluency.github.io/r-linear/r-linear-files.zip",
  destfile="r-linear-files.zip")
unzip("r-linear-files.zip")

# Install some CRAN packages:
install.packages(c(
    "tidyverse", "multcomp", "emmeans", 
    "lme4", "lmerTest", "pbkrtest", "BiocManager"))

# Install some Bioconductor packages:
BiocManager::install(c("limma","edgeR","topconfects"))
```

Now load the file `linear_models.R` in the `r-linear-files` folder.

## Files

* [r-linear-files.zip](r-linear-files.zip) - Files used in this workshop.

## Key functions to remember

Built-in to R: 

    lm, model.matrix, coef, sigma, df.residual,
    predict, confint, summary, anova, drop1, 
    I, poly

`splines` -- curve fitting:

    ns, bs

`multcomp` and `emmeans` -- linear hypothesis tests and multiple comparisons: 

    glht, mcp, confint, summary, emmeans

`limma` and `edgeR` -- fitting many models to gene expression data: 

    DGEList, calcNormFactors, cpm,
    lmFit, contrasts.fit, eBayes, plotSA, topTable

## Links

* Postgraduate students at Monash can access [statistical consulting](https://www.monash.edu/researchinfrastructure/datascienceandai/capabilities/statistical-consulting), courtesy of the Data Science and AI Platform. This is a good service for beginner to intermediate statistical questions.

* The [Biostatistics Consulting Platform](https://www.monash.edu/medicine/sphpm/units/biostats-consulting) in the Monash Faculty of Medicine may be more suitable for advanced questions about experimental design and analysis.

<br>

* [Monash Data Fluency](https://www.monash.edu/data-fluency)
* [Monash Bioinformatics Platform](https://www.monash.edu/researchinfrastructure/bioinformatics)
* [More workshop material from Monash Bioinformatics Platform](https://www.monash.edu/researchinfrastructure/bioinformatics/training)

<br>

* [Course notes for PH525x](http://genomicsclass.github.io/book/) (initial chapters of this edX course cover similar material to this workshop)
* [James, Witten, Hastie and Tibshirani (2013) "An Introduction to Statistical Learning"](https://www.statlearning.com/)
* [Dance of the CIs app](http://logarithmic.net/2017/dance/) (to give intuition about Confidence Intervals)
* [The Art of Linear Algebra](https://github.com/kenjihiranabe/The-Art-of-Linear-Algebra/blob/main/The-Art-of-Linear-Algebra.pdf) (to give intuition about matrices and vectors -- sections 1-3 are relevant to this workshop)
* Testing for differential gene expression often uses linear models. The folks at [WEHI](https://www.wehi.edu.au/research/research-fields/bioinformatics) have written good introductions to this topic:
    * ["RNA-seq analysis is easy as 1-2-3 with limma, Glimma and edgeR"](https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html)
    * ["A guide to creating design matrices for gene expression experiments"](https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html) (design matrix = model matrix)
* Mixed effects models are a popular next step beyond the fixed effects models covered in this workshop.
    * [Mixed Models in R](https://m-clark.github.io/mixed-models-with-R/)


## Author

This course has been developed for the [Monash Bioinformatics Platform](https://www.monash.edu/researchinfrastructure/bioinformatics) and [Monash Data Fluency](https://www.monash.edu/data-fluency) by Paul Harrison.

<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="figures/CC-BY.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.

## Source code

* [GitHub repository](https://github.com/MonashDataFluency/r-linear)

<p style="margin-top: 5em; text-align: right">
<a href="https://www.monash.edu/researchinfrastructure/bioinformatics"><img src="figures/MBP-banner.png" width="675"></a>
</p>
