---
title: "RNA-Seq differential expression analysis in R slideshow"
output: 
    ioslides_presentation:
        widescreen: true
        smaller: true
        css: style.css
        incremental: true
        mathjax: "https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"
---

## Screening many responses

"High throughput" biological datasets we can view as containing many sets of responses.

A typical example is RNA-Seq gene expression.

<img src="../figures/igv.png" height="450">


## RNA-Seq

> * Perform RNA-Seq on a set of biological samples.
> * Number of RNA molecules seen for each gene in each sample is counted.

> * Linear model formula based on experiment design.
> * Each gene can be viewed as a separate set of responses.

<br>

### Key features

> * A test is performed for each gene. Multiple testing adjustment is crucial!
> * Residual standard deviation is similar between genes. "Empirical Bayes" shrinking of these standard deviations overcomes problems with small numbers of samples.

<br>

Use Bioconductor packages such as `limma` (fit many linear models) or `edgeR` (fit many GLMs).


## Further resources

### From the authors of limma and edgeR at WEHI

> * ["RNA-seq analysis is easy as 1-2-3 with limma, Glimma and edgeR"](https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html)

> * ["A guide to creating design matrices for gene expression experiments"](https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html)

### From Monash

> * [MGBP's RNA-Seq end-to-end workshop](https://monashbioinformaticsplatform.github.io/RNAseq_workshop_2024/)

> * [Degust](https://degust.erc.monash.edu/) for web-based differential expression analysis
